rasa/graph_components/providers/nlu_training_data_provider.py

```diff
@@ -0,0 +1,87 @@
+from __future__ import annotations
+from typing import Dict, Text, Any, Optional
+import os
+from rasa.engine.graph import GraphComponent, ExecutionContext
+from rasa.engine.storage.resource import Resource
+from rasa.engine.storage.storage import ModelStorage
+from rasa.shared.importers.importer import TrainingDataImporter
+from rasa.shared.nlu.training_data.training_data import (
+    TrainingData,
+    DEFAULT_TRAINING_DATA_OUTPUT_PATH,
+)
+from rasa.shared.nlu.training_data.loading import load_data
+
+
+class NLUTrainingDataProvider(GraphComponent):
+    """Provides NLU training data during training and inference time."""
+
+    def __init__(
+        self,
+        model_storage: ModelStorage,
+        resource: Resource,
+        training_data: Optional[TrainingData] = None,
+    ) -> None:
+        """Creates NLU training data provider."""
+        self._model_storage = model_storage
+        self._resource = resource
+        self._training_data = training_data
+
+    @classmethod
+    def get_default_config(cls) -> Dict[Text, Any]:
+        return {"language": "en", "persist": False}
+
+    @classmethod
+    def create(
+        cls,
+        config: Dict[Text, Any],
+        model_storage: ModelStorage,
+        resource: Resource,
+        execution_context: ExecutionContext,
+    ) -> NLUTrainingDataProvider:
+        """Creates component (see parent class for full docstring)."""
+        return cls(model_storage, resource)
+
+    @classmethod
+    def load(
+        cls,
+        config: Dict[Text, Any],
+        model_storage: ModelStorage,
+        resource: Resource,
+        execution_context: ExecutionContext,
+        **kwargs: Any,
+    ) -> NLUTrainingDataProvider:
+        """Creates provider using a persisted data."""
+        with model_storage.read_from(resource) as resource_directory:
+            training_data = load_data(
+                resource_name=os.path.join(
+                    str(resource_directory), DEFAULT_TRAINING_DATA_OUTPUT_PATH
+                ),
+                language=config["language"],
+            )
+        return cls(model_storage, resource, training_data)
+
+    def _persist(self, training_data: TrainingData) -> None:
+        """Persists NLU training data to model storage."""
+        with self._model_storage.write_to(self._resource) as resource_directory:
+            training_data.persist(
+                dir_name=str(resource_directory),
+                filename=DEFAULT_TRAINING_DATA_OUTPUT_PATH,
+            )
+
+    def provide(
+        self, config: Dict[Text, Any], importer: Optional[TrainingDataImporter] = None,
+    ) -> TrainingData:
+        """Provides nlu training data during training."""
+
+        if config == {}:  # use default config if config is empty
+            config = self.get_default_config()
```

Comment:
You can move this to the constructor and there use something like `self._config = {**self.default_config(), **config}` which will fill the `self._config` with the default parameters first and then overwrite default values with the ones from the `config` passed to the constructor

Reply:
The `GraphNode` which which wraps/instantiates the `GraphComponent` will take care of this for us :) (see: [here](https://github.com/RasaHQ/rasa/blob/5bb331ff7ba75ad564478cbc71fd1a46bce5ee02/rasa/engine/graph.py#L315))